modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/DelegatingStorageRetryStrategy.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.repositories.gcs;
+
+import com.google.api.gax.retrying.ResultRetryAlgorithm;
+import com.google.api.gax.retrying.TimedAttemptSettings;
+import com.google.cloud.storage.StorageRetryStrategy;
+
+import java.util.concurrent.CancellationException;
+import java.util.function.Function;
+
+public class DelegatingStorageRetryStrategy<T> implements StorageRetryStrategy {
+
+    private final ResultRetryAlgorithm<T> idempotentRetryAlgorithm;
+    private final ResultRetryAlgorithm<T> nonIdempotentRetryAlgorithm;
+
+    /**
+     * @param delegate The delegate {@link StorageRetryStrategy}
+     * @param decorator The function applied to both idempotent and non-idempotent {@link ResultRetryAlgorithm}s
+     */
+    @SuppressWarnings("unchecked")
+    public DelegatingStorageRetryStrategy(
+        StorageRetryStrategy delegate,
+        Function<ResultRetryAlgorithm<T>, ResultRetryAlgorithm<T>> decorator
+    ) {
+        this.idempotentRetryAlgorithm = decorator.apply((ResultRetryAlgorithm<T>) delegate.getIdempotentHandler());
+        this.nonIdempotentRetryAlgorithm = decorator.apply((ResultRetryAlgorithm<T>) delegate.getIdempotentHandler());
```

Comment:
```suggestion
        this.nonIdempotentRetryAlgorithm = decorator.apply((ResultRetryAlgorithm<T>) delegate.getNonidempotentHandler());
```

It does not really matter in practice since the legacy retry strategy use the same algorithm for both handlers. But I wonder whether this is theoretically more correct and does not depend on the implementation details?